R/req-perform-stream.R

```diff
@@ -214,39 +209,116 @@ resp_stream_lines <- function(resp, lines = 1) {
   readLines(conn, n = lines)
 }
 
+# Slices the vector using the only sane semantics: start inclusive, end
+# exclusive.
+#
+# * Allows start == end, which means return no elements.
+# * Allows start == length(vector) + 1, which means return no elements.
+# * Allows zero-length vectors.
+#
+# Otherwise, slice() is quite strict about what it allows start/end to be: no
+# negatives, no reversed order.
+slice <- function(vector, start = 1, end = length(vector) + 1) {
+  stopifnot(start > 0)
+  stopifnot(start <= length(vector) + 1)
+  stopifnot(end > 0)
+  stopifnot(end <= length(vector) + 1)
+  stopifnot(end >= start)
+
+  if (start == end) {
+    vector[FALSE] # Return an empty vector of the same type
+  } else {
+    vector[start:(end - 1)]
+  }
+}
+
+# Function to find the first double line ending in a buffer, or NULL if no
+# double line ending is found
+#
+# Example:
+#   find_event_boundary(charToRaw("data: 1\n\nid: 12345"))
+# Returns:
+#   list(
+#     matched = charToRaw("data: 1\n\n"),
+#     remaining = charToRaw("id: 12345")
+#   )
+find_event_boundary <- function(buffer) {
+  if (length(buffer) < 2) {
+    return(NULL)
+  }
+
+  # leftX means look behind by X bytes. For example, left1[2] equals buffer[1].
+  # Any attempt to read past the beginning of the buffer results in 0x00.
+  left1 <- c(0x00, head(buffer, -1))
+  left2 <- c(0x00, head(left1, -1))
+  left3 <- c(0x00, head(left2, -1))
+
+  boundary_end <- which(
+    (left1 == 0x0A & buffer == 0x0A) | # \n\n
+    (left1 == 0x0D & buffer == 0x0D) | # \r\r
+    (left3 == 0x0D & left2 == 0x0A & left1 == 0x0D & buffer == 0x0A) # \r\n\r\n
+  )
+  
+  if (length(boundary_end) == 0) {
+    return(NULL)  # No event boundary found
+  }
+
+  boundary_end <- boundary_end[1]  # Take the first occurrence
+  split_at <- boundary_end + 1  # Split at one after the boundary
+
+  # Return a list with the event data and the remaining buffer
+  list(
+    matched = slice(buffer, end = split_at),
+    remaining = slice(buffer, start = split_at)
+  )
+}
+
 #' @export
 #' @rdname resp_stream_raw
 # TODO: max_size
```

Comment:
Maybe we can implement `max_size()` now that we have `buffer`. But what should happen if `max_size` is reached? Error? Return `NULL`? Some other sentinel?

Reply:
Error for sure, IMHO. I'll add that

Reply:
Now implemented